Add SSO

Simon-Laux opened the pull request:
closes #1

**TODO**:

step 1:

- [x] make SSO work
- [x] patch immich load client secret from file
- [x] make a flake for patched immich to avoid rebuilding every time even when nothing changed 
- [x] default on option to disable password login and disable the manual admin creation on first setup
   - [x] disable normal login
   - [x] block admin creation page in nginx
   - [x] make a startup systemd service/script that creates admin account via immich api on startup (if it is not set)
   - [x] and confirm that first visitor after installation can't create admin account anymore

---
step 2:

- [ ] bring back `cfg.defaultStorageClaim` option (either by implementing it as string field and converting in nixos or by fixing the bug with int fields in SP)
- [ ] make a proper upstream pr to immich to allow loading client secret from file. 
- [ ] immich admin permission should apply from SP to immich (does not work yet)
    - [ ] is it automatically updated on login when a user is promoted or demoted? if not we may need some script or workaround to handle this

Simon-Laux committed:
start with sso

Simon-Laux committed:
try fix nix issue

Simon-Laux committed:
try different fix approach

Simon-Laux committed:
do the simple solution, not as pretty but should work

Simon-Laux committed:
fix assertion

Simon-Laux committed:
fix value type

thenhnn commented on module.nix:
```diff
@@ -111,5 +147,79 @@ in
+      buttonText = "Login with Kanidm";
+
+      clientId = "immich";
+      clientSecret = oauthClientSecretFP;
```
Are you sure Immich will read client secret from file?

Simon-Laux replied:
good catch. seems like it can not do it yet. I'll try `builtins.readFile oauthClientSecretFP`, but the ideal solution would be to make a pr to immich to add the feature.

thenhnn replied:
I think you can inline patch in the repo. Fetching it from GitHub seems to only complicate things :)

Simon-Laux committed:
add different weight

thenhnn commented on module.nix:
```diff
@@ -111,5 +147,79 @@ in
+      # storageQuotaClaim = "immich_quota";
+    };
+
+    # TODO redirect uris
```
I'm not sure if Kanidm even allows multiple redirect uris per client, so it might not be possible to add additional redirect uri for immich mobile app :(

thenhnn replied:
Kanidm supports that.

https://git.selfprivacy.org/SelfPrivacy/selfprivacy-nixos-config/pulls/173 allows to pass multiple values to ` selfprivacy.auth.clients.<name>.originUrl`.

thenhnn commented on config-paths-needed.json:
```diff
@@ -2,5 +2,14 @@
-  ["selfprivacy", "modules", "immich"]
+  ["selfprivacy", "modules", "immich"],
+  ["selfprivacy", "modules", "auth", "enable"],
+  ["selfprivacy", "passthru", "auth", "ldap-base-dn"],
```
I don't think you need LDAP stuff here.

thenhnn commented on module.nix:
```diff
@@ -111,5 +147,79 @@ in
+      enabled = true;
+      autoRegister = true;
+      # https://immich.app/docs/administration/oauth/#auto-launch
+      autoLaunch = false;
```
Is there any reason to disable `autoLaunch`, but rewrite /user/login to /user/oauth2 in NGINX? Seems like both do the same thing.

Simon-Laux replied:
just copied carelessly, I'll enable it here instead of doing the nginx rewrite.

Simon-Laux committed:
remove non sensical redirect and add other redirect uris

Simon-Laux committed:
remove ldap stuff

Simon-Laux committed:
read client secret from file

Simon-Laux committed:
patch immach to support loading oauth clientSecret from file

Simon-Laux committed:
no hash

Simon-Laux committed:
correct hash

Simon-Laux committed:
trust Pkce will work and not set client secret

Simon-Laux committed:
enable Pkce mode that does not use the client secret

Simon-Laux committed:
delete client secret & improve originLanding url so that login process
starts directly

Simon-Laux committed:
fix nix error

Simon-Laux committed:
fix typo

Simon-Laux committed:
remove merge which is incorrect at this place

Simon-Laux committed:
set signingAlgorithm to ES256

Simon-Laux committed:
temporaly hardcode defaultStorageClaim to check if it is responsible
that the services screen doesn't load anymore

Simon-Laux committed:
fix temp rm defaultStorageClaim option in hope that this fixes the
selfprivacy ui

Simon-Laux committed:
disable password login

thenhnn commented on module.nix:
```diff
@@ -111,5 +150,89 @@ in
+      };
+    };
+
+    # enable Pkce mode that does not use the client secret
```
Confidential mode (the only one supported by Immich according to the docs) will still require client secret, even with PKCE.

thenhnn replied:
https://github.com/immich-app/immich/discussions/4459#discussioncomment-13184960

Simon-Laux replied:
yesterday I didn't manage to make a patched version of immich because the package has some quirks that made it a bit harder to patch (src is not easily overwritable, because it is [inside of a let variable](https://github.com/NixOS/nixpkgs/blob/d179d77c139e0a3f5c416477f7747e9d6b7ec315/pkgs/by-name/im/immich/package.nix#L107) and the source structure was a bit different, so I needed to [patch my patch](https://github.com/Simon-Laux/immich-selfprivacy-module/blob/5fb69d0c5a649d8108b953dc594e002735fdd0cb/patched-immich/flake.nix#L16-L35)). but today I managed it and reintroduced the client secret.

But eventually we need to make a proper pr to immich to add the feature of "loading client secret from file" to upstream – alone for the fact that maintaining patches across newer versions will be bothersome.

Simon-Laux committed:
reintroduce client secret with patched immich

Simon-Laux committed:
try making a seperate flake for the patched immich to avoid recompiling
it every time

Simon-Laux committed:
try another method

Simon-Laux committed:
try again

Simon-Laux committed:
try fix

Simon-Laux committed:
try fix again

Simon-Laux committed:
remove commented out code that removes client secret

Simon-Laux committed:
use `pkgs.system` instead of `builtins.currentSystem`

Simon-Laux committed:
start exploration to prevent first time admin logins screen

Simon-Laux committed:
also restrict the outside access to the api

Simon-Laux committed:
fix order of nginx locations

Simon-Laux committed:
try fix

Simon-Laux committed:
comment out "/auth/register" blocking rule since it does not work anyway

Simon-Laux committed:
add option to only allow SSO Login and auto register the admin account

Simon-Laux committed:
fix slice error

Simon-Laux committed:
passwordLogin should follow cfg.OnlyAllowSSOLogin option

Simon-Laux committed:
added more verbose logging to script to find a bug

Simon-Laux committed:
comment out "requires" in immich-auto-register-admin service, in hope
that it fixes the startup issue on switching `cfg.OnlyAllowSSOLogin`
from the off to the on state

Simon-Laux committed:
find and fix the statup issue of `immich-auto-register-admin`

Simon-Laux committed:
update readme
